### DISCUSSION TITLE
Enhance focus styles

### DISCUSSION THREAD

**Discussion by basham** [indiana-university/rivet-source#552] [Feature and enhancement requests]
Rivet uses a 2px blue outline for a focus style. This is fine for typical cases when the focusable content is on a white background. But I want to explore how a focus style could work on a dark or red background, and if a single focus style could work in these wider range of contexts.

There's a number of new resources available that explains how to design accessible focus styles:

1. [A guide to designing accessible, WCAG-compliant focus indicators](https://www.sarasoueidan.com/blog/focus-indicators/)
2. [Understanding Success Criterion 2.4.11: Focus Appearance (Minimum)](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum.html)
3. [Creating a two-color focus indicator to ensure sufficient contrast with all components](https://www.w3.org/WAI/WCAG21/Techniques/css/C40)
4. [Using an author-supplied, highly visible focus indicator](https://www.w3.org/WAI/WCAG21/Techniques/general/G195)

And we have a number of places that already use alternative styling:

1. Red buttons with red focus style. White buttons with white focus style. https://framework.iu.edu/
2. Red footer with white focus style. https://www.iu.edu/brand/brand-expression/visual-language/index.html
3. Red footer with blue focus style. https://studios.iu.edu/about/people/brand-teams/index.html

These CodePens explore some alternative directions, using Rivet 1.8.1. They are all on a gray background. Future work would be to explore using red and dark backgrounds.

1. Standard focus style. https://codepen.io/basham/pen/QWgENvN
2. Experiment 1: Dark-light outset focus style. https://codepen.io/basham/pen/xxdbGov
3. Experiment 2: Dark-light-dark inset focus style. https://codepen.io/basham/pen/QWvwyGL
4. Experiment 3: Dark-light-dark outset focus style. https://codepen.io/basham/pen/QWvwyxj

By the way, these experiments were shared with the design community in September. The general consensus among the three experiments is to go with Experiment 3. That of course doesn't preclude us exploring even more options.

Maybe a multicolor focus style is not worth the complexity. Maybe all we need is a simple way to change the focus color for the context. Maybe this is a non-issue. Don't know.

**Comment by levimcg** [comment]
@basham Thanks for writing this up and creating these test cases. We are doing some planning right now for the future beta releases and would like to move forward with making some updates to focus styles based on these recommendations.

Quick question. In _Experiment 1_ the slightly more subtle focus ring seems really appealing. Does it meet the WCAG success criterion for contrast? Just wanted to make sure before we decide on a final direction. Thanks!

**Reply by basham** [reply]
After studying the spec some more, I would say, Experiment 1 passes the test for links but not inputs.

For links, the 1px blue outline is sufficient contrast with the standard white backgrounds. Because it is an outline, it has sufficient size. The blue-200 (`#b8c8dc`) would be considered only aesthetic, because it doesn't provide enough contrast against a white background.

For inputs, the border color changes from gray (`#666666`) to blue-700 (`#134a71`). This change does not pass the 3:1 contrast requirements. The focus color would need to be almost black to minimally meet the contrast against the gray. And again, the blue-200 doesn't provide enough contrast against white to provide a benefit.

However, I'd say Experiment 2 passes. The border color change and the blue-200 provides no benefit to the contrast area. However, the outermost 1px blue-700 (that is the change compared to Experiment 1) is the differentiator. Blue-700 has a 9.35:1 contrast against a white background. Essentially, we could choose to include only that part of the focus indicator, and it would pass. But by not having that part, none of it works.

I do think there's more options to explore than these three experiments. If you have a particular aesthetic in mind, let's discuss.

**Reply by basham** [reply]
I'm also interested in experimenting with how the look and contrast works with the [new Rivet 2 colors](https://v2.rivet.iu.edu/docs/utilities/color/). I'm assuming you'd want to use those colors, anyway.

**Comment by basham** [comment]
This recent article discusses many of the same concerns brought up in this work. It may be a good reference as we explore different options.

[*Developing a focus style for a themable design system*](https://adhoc.team/2022/02/08/creating-focus-style-for-themable-design-system/)
